components/calendar/src/dangi.rs

```diff
@@ -0,0 +1,1238 @@
+    /// assert_eq!(dangi_datetime.time.minute.number(), 1);
+    /// assert_eq!(dangi_datetime.time.second.number(), 0);
+    /// ```
+    pub fn try_new_dangi_datetime(
+        year: i32,
+        month: u8,
+        day: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+    ) -> Result<DateTime<Dangi>, CalendarError> {
+        Ok(DateTime {
+            date: Date::try_new_dangi_date(year, month, day)?,
+            time: types::Time::try_new(hour, minute, second, 0)?,
+        })
+    }
+}
+
+impl ChineseBased for Dangi {
+    fn location(fixed: RataDie) -> Location {
+        if fixed < FIXED_1908 {
+            KOREAN_LOCATION_ORIGINAL
+        } else if fixed < FIXED_1912 {
+            KOREAN_LOCATION_1908
+        } else if fixed < FIXED_1954 {
+            KOREAN_LOCATION_1912
+        } else if fixed < FIXED_1961 {
+            KOREAN_LOCATION_1954
+        } else {
+            KOREAN_LOCATION_1961
+        }
+    }
+
+    const EPOCH: RataDie = KOREAN_EPOCH;
+
+    // Unwrap can be used for this function because it is only called internally when
+    // generating dates from a valid year, month, and day
+    #[allow(clippy::unwrap_used)]
+    fn new_chinese_based_date(year: i32, month: u8, day: u8) -> ChineseBasedDateInner<Dangi> {
+        ChineseBasedDateInner(ArithmeticDate::new_unchecked(year, month, day))
+    }
+}
+
+impl Dangi {
+    /// Get the ISO Date of Seollal (Korean Lunar New Year) on or before the given ISO Date `iso`.
+    ///
+    /// ```rust
+    /// use icu::calendar::Date;
+    /// use icu::calendar::dangi::Dangi;
+    ///
+    /// let iso = Date::try_new_iso_date(2023, 6, 18)
+    ///     .expect("Failed to initialize ISO Date instance.");
+    /// let seollal = Dangi::seollal_on_or_before_iso(iso);
+    ///
+    /// assert_eq!(seollal.year().number, 2023);
+    /// assert_eq!(seollal.month().ordinal, 1);
+    /// assert_eq!(seollal.day_of_month().0, 22);
+    /// ```
+    pub fn seollal_on_or_before_iso(iso: Date<Iso>) -> Date<Iso> {
+        let iso_inner = iso.inner;
+        let fixed = Iso::fixed_from_iso(iso_inner);
+        let result_fixed = Inner::new_year_on_or_before_fixed_date(fixed);
+        Iso::iso_from_fixed(result_fixed)
+    }
+
+    fn format_dangi_year(year: i32) -> FormattableYear {
+        let era = Era(tinystr!(16, "dangi"));
+        let number = year;
+        let cyclic = Some(div_rem_euclid(number - 1 + 364, 60).1 + 1);
+        let mid_year = Inner::fixed_mid_year_from_year(number);
+        let iso_formattable_year = Iso::iso_from_fixed(mid_year).year();
+        let related_iso = Some(iso_formattable_year.number);
+        types::FormattableYear {
+            era,
+            number,
+            cyclic,
+            related_iso,
+        }
+    }
+
+    /// Get the ordinal lunar month from a code
+    ///
+    /// TODO: This code is exactly the same as in `Chinese`; consider abstracting this function
+    /// to a trait shared by Lunar calendars.
```

Comment:
Suggestion: it doesn't need to be a trait function. It can just be a free function imported into both places.

Reply:
Done